removes Draggable's dependencies on its parent component

shaoshing opened the pull request:
This PR removes the Draggable component's dependencies on its parent component to pass in the **dragPosition** and **isDragTarget** by retrieving them from the DragAndDrop store directly. This will avoid duplicate in the LibrariesPanel and Tab components. More importantly, it makes the code easier to understand. 

##### Bonus 

Also resolved a position bug when you drag and move a layer fast.
**before** ![drag-before](https://cloud.githubusercontent.com/assets/118264/8881715/b02b21bc-31f6-11e5-948d-66ddbccbe520.gif)

**after** ![drag-after](https://cloud.githubusercontent.com/assets/118264/8881728/c968c51c-31f6-11e5-9400-a4057a5c23d9.gif)

##### Thoughts about future top-down refactoring

I against the way of wrapping the target component with higher level components (e.g. Draggable/Droppable, or  the similar [react-dnd](https://github.com/gaearon/react-dnd) package), as it tends to make the data flow and communication vague:

* higher level component (e.g. Draggable/Droppable) hides the props required by the target component (e.g. LayerFace). Basically we have to pass the props to Droppable/Draggable and then use {...this.state.props} to pass down to LayerFace. This tend to be buggy.

* the communications between the components are unclear. 

* configuration is difficult and confusing.

A better way might be wrapping the content of the component, instead of the component. Here is a [gist](https://gist.github.com/shaoshing/16a4fd0e2fb27a06d61c#file-example-jsx) I created as a demonstration of the Draggable component. The gist does not include every detail, but if you just look at the Item component, you have a much better understanding of the data flow and communication, and it is also very React style. The trick is to the `React.cloneElement(this.props.children[0], { ref: dragTarget })`, which allows us encapsulate the event listening inside the Draggable component, leaving the Item component clean and concise.

@iwehrman @designedbyscience  Please review and let me know you thoughts.

designedbyscience commented:
@shaoshing You should use Vermillion to check this...It seems much slower than the current drag and drop (intolerably so)

designedbyscience commented:
For your proposed refactoring: I like the sound of that, I hadn't looked into the cloneElement trick

shaoshing commented:
Hey @designedbyscience , the performance issue is fixed by manually listening to the DragAndDrop store such that when dragging, there is only one layer (instead of all layer) that is responding to the drag updates. Now it is as fast as the original one :dancers:

designedbyscience commented:
Okay, now 1 bug with layers not dropping correctly and we can't drag multiple items at once

iwehrman commented:
I really like where this is going! But, before I dig further, I'd like to point out that there is still a very hot spot in the profile that happens when dragging around the area where the drag started:

![drop](https://s3.amazonaws.com/f.cl.ly/items/443b1W451U1F1P3R1W3t/Image%202015-07-25%20at%2011.36.46%20AM.png)

In that profile above, those areas of intense computation are happening when I drag the layers around where I originally picked them up. Performance otherwise seems fine, but the rendering delays there are huge. Can you look into that? Also, as @designedbyscience suggested, testing with VermilionArtboards.psd is essential. Also, you should expand _all_ the layers when doing this testing. We're still planning to limit the total number of layers present in the DOM at any time (updating the set on scroll, for example), but I want to keep the core rendering as efficient as possible in case we run into trouble with that plan.

shaoshing commented:
The reason of the peaks is  because the `isDropTarget` property changed from `false -> null ` (`null -> false`) and thus triggers all LayerFace re-rendered when you drag in to (out of) the original positions of the dragged layers. The fix is making sure to convert `null` to `false` before passing to LayerFace (https://github.com/adobe-photoshop/spaces-design/pull/1829/files#diff-599bf64b66bfd695d31114df0d8d0c63R554). (Note that the master branch has the same issue).

![after](https://cloud.githubusercontent.com/assets/118264/8936897/d4f20f4a-3509-11e5-871d-9e027cff53f1.png)

**Another note for future refactoring**

Since the LayerFace relies on the LayerPanel heavily, each re-render of the LayerPanel will trigger all LayerFace to re-render. Even though we are using `shouldComponentUpdate` to avoid re-rendering in the children, but the total time of all function calls (in React) could be a lot. We've removed the dependencies in Draggable, so I went a little bit further by removing the dependencies in the Droppable, and the result is 40% faster:

 ![droppabel-before](https://cloud.githubusercontent.com/assets/118264/8937095/288a5076-350b-11e5-80fc-2953669daf1e.png) Before 
 ![droppable-after](https://cloud.githubusercontent.com/assets/118264/8937097/2d65daf2-350b-11e5-87f6-554836bf062f.png) After. Most of the `drag move` are below the 60FPS line, 

To make the Droppable independent, we need to use `onMouseEnter` and `onMouseLeave` to know which Droppable is the drop target, and also setting css `pointer-events: none;` to the dragged item to allow elements below it to receive mouse events.

@iwehrman  @designedbyscience  please review.

iwehrman commented on src/js/jsx/sections/layers/LayersPanel.jsx:
```diff
@@ -467,7 +467,7 @@ define(function (require, exports, module) {
                 var dragSource = collection.pluck(this.state.dragTargets, "id");
 
-                flux.actions.layers.reorder(doc, dragSource, dropIndex)
+                return flux.actions.layers.reorder(doc, dragSource, dropIndex)
```
I know you didn't add this `then` handler, but it's a little odd that it's in the JSX instead of as part of some other action. I only mention it because there is not actually any reason why this method needs to return a promise.

shaoshing replied:
The reason behind this is that the "dragging" state should maintain until the drop process is finished. Without this, the dropped layer will snap back after dropping, or you have to add information into the draganddrop store to represents a `dropping` state. I think the promise way is a simpler and does not affect user experience.  I also added the reason here https://github.com/adobe-photoshop/spaces-design/pull/1829/files#diff-9adf1065d9513443f0903ee1061b1b0dR182

iwehrman replied:
OK, I understand a little better. I agree that the `finally` handler makes sense here, but I still think the `then` handler should be part of the action.

shaoshing replied:
the `then` part is from the master, but I agree with you on that. Let me see if I can move to the action safely.

iwehrman commented on src/js/jsx/sections/layers/LayersPanel.jsx:
```diff
@@ -550,8 +551,7 @@ define(function (require, exports, module) {
                     .map(function (layer, visibleIndex) {
-                        var isDragTarget = dragTargets && dragTargetSet.has(layer),
-                            isDropTarget = dropTarget && dropTarget.key === layer.key,
+                        var isDropTarget = !!dropTarget && dropTarget.key === layer.key,
```
This isn't really a nit, but I don't think it's actually necessary to convert the first parameter into a boolean here.

shaoshing replied:
Adding the `!!` will make sure `isDropTarget` is always a boolean value, this is the cause of the performance issue when dragging over dragged items. So, before you drag over the dragged items, the `dropTarget` has value and `isDropTarget ` is false for all non-drop targets, but when you drag over the dragged items, `dropTarget` changed to null, so as the `isDropTarget`, and cause all non-drop targets to do a re-render because `this.props.isDropTarget !== nextProps.isDropTarget`.

iwehrman replied:
Doh, yes, I get it.

iwehrman commented on src/js/jsx/shared/Draggable.jsx:
```diff
@@ -99,71 +117,77 @@ define(function (require, exports, module) {
                 };
             },
+            
+            // Instead of responding to DragAndDrop's all change events, it listens to the "start-drag" event (fires 
```
For consistency, methods and functions should be prefixed with block-style comments instead of line-style comments. Also, please mark it as `@private`.

shaoshing replied:
Got it.

iwehrman commented:
This is tremendous. Thank you for helping me understand your changes. I still have a few comment nits and possibly would like to push a little more of the drop handler back into reorder (if that's appropriate), but once that's cleaned up I'm ready to merge.

shaoshing commented:
Thanks for reviewing @iwehrman  , I updated the function doc and moved some drop handling code to where it belongs to (actions.layers). Now back to you.

iwehrman commented:
Looks good!

designedbyscience commented:
Awesome!

chadrolfs commented:
ahhhh, much better
